src/python-common/ceph/rgw/rgwam_core.py

```diff
@@ -612,99 +668,158 @@ def realm_rm_zone_creds(self, realm_token_b64):
-        #    ep = ''
-        #    if len(eps) > 0:
-        #        ep = eps[0]
-        #        o = urlparse(ep)
-        #        port = o.port
-        #        spec = RGWSpec(service_id = svc_id,
-        #                       rgw_realm = realm_name,
-        #                       rgw_zone = zone_name,
-        #                       rgw_frontend_port = o.port)
-        #        self.env.mgr.apply_rgw(spec)
+        realm = EntityID(realm_id)
+        zonegroup = self.create_zonegroup(realm, zonegroup_name, zonegroup_is_master, endpoints)
+        self.update_period(realm, zonegroup)
 
-        self.env.mgr.apply_rgw(svc_id, realm_name, zone_name)
+        return (0, f'Created zonegroup {zonegroup_name} on realm {realm.name}', '')
 
-        daemons = self.env.mgr.list_daemons(svc_id, 'rgw', refresh=True)
+    def zone_create(self, rgw_spec, start_radosgw):
 
-        ep = []
-        for s in daemons:
-            for p in s.ports:
-                ep.append('http://%s:%d' % (s.hostname, p))
+        if not rgw_spec.rgw_realm_token:
+            raise RGWAMException('missing realm token')
+        if rgw_spec.rgw_zone is None:
+            raise RGWAMException('Zone name is a mandatory parameter')
+        if rgw_spec.rgw_zone in self.zone_op().list():
+            raise RGWAMException(f'Zone {rgw_spec.rgw_zone} already exists')
 
-        log.error('ERROR: ep=%s' % ','.join(ep))
+        realm_token = RealmToken.from_base64_str(rgw_spec.rgw_realm_token)
+        if realm_token.endpoint is None:
+            raise RGWAMException('Provided realm token has no endpoint')
 
+        access_key = realm_token.access_key
+        secret = realm_token.secret
         try:
-            zone_info = self.zone_op().modify(zone, zg, endpoints=','.join(ep))
+            realm_info = self.realm_op().pull(EntityName(realm_token.realm_name),
+                                              realm_token.endpoint, access_key, secret)
         except RGWAMException as e:
-            raise RGWAMException('failed to modify zone', e)
+            raise RGWAMException('failed to pull realm', e)
 
-        return (0, success_message, '')
+        logging.info(f"Pulled realm {realm_info['name']} ({realm_info['id']})")
+        realm_name = realm_info['name']
+        realm_id = realm_info['id']
+
+        realm = EntityID(realm_id)
+        period_info = self.period_op().get(realm)
+        period = RGWPeriod(period_info)
+        logging.info('Period: ' + period.id)
+
+        zonegroup = period.find_zonegroup_by_name(rgw_spec.rgw_zonegroup)
+        if not zonegroup:
+            raise RGWAMException(f'zonegroup {rgw_spec.rgw_zonegroup}')
+
+        zone = self.create_zone(realm, zonegroup, rgw_spec.rgw_zone, False, access_key, secret)
+        self.update_period(realm, zonegroup, zone)
+
+        period = RGWPeriod(period_info)
+        logging.debug(period.to_json())
+
+        if start_radosgw:
+            secondary_realm_token = RealmToken(realm_name,
```

Comment:
What's the point of the `secondary_realm_token`? You won't be pulling from this zone if you were setting up another cluster.

Reply:
@alimaredia it's used to call back the rgw module from cephadm (to update the end-points once the daemons are created). That's the only purpose of it.

Reply:
By update the endpoints do you mean update the endpoints for the token or to update the endpoints of the rgw daemons that are stored internally somehow? 

The Ceph cluster admin deploying multisite would never use these secondary_realm_tokens, so can we get rid of them to avoid confusion?

Reply:
@alimaredia the token in this case is used `internally` (by cephadm) to update the endpoints of the rgw secondary zone by using the info from the cephadm daemons. 

Anyway, in general the tokens are not stored: they are generated on the fly by using the current real information. So if a user runs `ceph rgw realm tokens` they should get up-to-date tokens for each realm.

Reply:
Just to be clear the up-to-date tokens for each realm have endpoints in the master zone right?

Reply:
> Just to be clear the up-to-date tokens for each realm have endpoints in the master zone right?

Yes, we always look for the master zone of the realm and use its endpoints when generating the tokens.